### PR TITLE
One hook to rule them all: preparation for a block supports API

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

This PR refactors to hooks to use a single filter for `editor.BlockEdit`, and with that lays the foundations for a new block supports API (the other filters remain to be done in a similar fashion).

## Why?

Related to https://github.com/WordPress/gutenberg/discussions/51005.

Less code duplication.
Less higher order components and less components in a block tree.
Less opportunities for performance regressions:
* Check if controls should be rendered in a single place.
* Make the components pure in a single place.

### CODE REVIEW

File: packages/block-editor/src/hooks/align.js [hunk 1]
```diff
@@ -152,45 +152,13 @@ function BlockEditAlignmentToolbarControlsPure( {
...
-						align={ props.attributes.align }
-						setAttributes={ props.setAttributes }
-					/>
-				) }
-				<BlockEdit key="edit" { ...props } />
-			</>
-		);
+export default {
+	edit: BlockEditAlignmentToolbarControlsPure,
+	attributeKeys: [ 'align' ],
```
Comment (ellatrix) [inline packages/block-editor/src/hooks/align.js:158]: @youknowriad I'm using this to "subscribe" to specific attributes only. But on the other hand, extra `useSelect` subscription these control components are not bad per se because they are only rendered when a block is selected. If we allow that, we wouldn't need to do this declaration of which attributes need to be passed down. Either way, I'm fine with it, they're not mutually exclusive.

Comment (ellatrix) [reply packages/block-editor/src/hooks/align.js:158]: And we can decide this later when we actually make the real API.

File: packages/block-editor/src/hooks/position.js [hunk 2]
```diff
@@ -318,44 +314,19 @@ export function PositionPanelPure( {
...
-export const withPositionControls = createHigherOrderComponent(
-	( BlockEdit ) => ( props ) => {
-		const { name: blockName } = props;
-		const positionSupport = hasBlockSupport(
-			blockName,
-			POSITION_SUPPORT_KEY
-		);
+export default {
+	edit: function Edit( props ) {
 		const isPositionDisabled = useIsPositionDisabled( props );
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/position.js:319]: This one is interesting, I feel like maybe we need one sync `hasSupport` key and potentially another `useHasSupport` key that can use hooks. (if this is more common)

**Review by youknowriad — COMMENTED** [review]
This is really cool. I like where this is going. Let's obviously keep that kind of "API" private but it clearly shows that we're repeating the same things over and over and that we can optimize these. 

Also I believe ultimately, that we should avoid using the filters entirely and just bundle these in the framework directly.

**Comment by youknowriad** [comment]
@Mamaduka @gziolo you're going to like this.

File: packages/block-editor/src/hooks/align.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-editor/src/hooks/align.js:158]: Yes, this is totally fine for me and to be honest, I'm not entirely sold on opening this API to the public yet but it's definitely a great improvement for us.

File: packages/block-editor/src/hooks/position.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/hooks/position.js:319]: We can alway add that later if needed. So far it's only one of them.

**Comment by ellatrix** [comment]
> Also I believe ultimately, that we should avoid using the filters entirely and just bundle these in the framework directly.

Yes, eventually, we should create a store and mount this component straight in the right place instead of with filters.

**Review by youknowriad — APPROVED** [review]
To be honest, this looks great as is. We can absorb more filters as we go. Curious if it has any impact on the performance. Not certain.

File: packages/block-editor/src/hooks/align.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/hooks/align.js:158]: Oh sure, no public API for now :D

**Comment by ellatrix** [comment]
@youknowriad Probably not, it's just easier to oversee everything and not make performance regressions.

**Comment by gziolo** [comment]
> To be honest, this looks great as is. We can absorb more filters as we go. Curious if it has any impact on the performance. Not certain.

That is exactly my thinking about the perf implications since it allows getting rid of at least a single higher-order component per feature!

> @Mamaduka @gziolo you're going to like this.

I couldn't agree more! Excellent refactor, @ellatrix! We don't even have to commit to any public interface for now. It's going to be discovered at the end of applying all necessary changes.

File: packages/block-editor/src/hooks/align.js [hunk 3]
```diff
@@ -153,6 +153,7 @@ function BlockEditAlignmentToolbarControlsPure( {
 }
 
 export default {
+	shareWithChildBlocks: true,
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/align.js:156]: I think I'd have preferred to leave this one to its own PR, to have a clear picture of impact.

Comment (ellatrix) [reply packages/block-editor/src/hooks/align.js:156]: It's needed, otherwise the controls are not rendered. For example for a button block, the align controls from the buttons blocks should be rendered.

**Comment by ramonjd** [comment]
This is awesome, thank you!

File: packages/block-editor/src/hooks/block-hooks.js
Comment (ockham) [inline packages/block-editor/src/hooks/block-hooks.js:1]: 👋 I might be missing some context, so I'm not sure it matters, but I just came to say that Block Hooks are not using the Block Supports mechanism; they're separate.

Comment (ellatrix) [reply packages/block-editor/src/hooks/block-hooks.js:1]: It doesn't matter, we can just turn on support for all blocks. The point is that controls are managed in one place and only rendered and re-rendered when necessary. In this case, since the controls don't depend on any attributes, we can prevent unnecessary re-renders.

File: packages/block-editor/src/hooks/block-hooks.js [hunk 4]
```diff
@@ -83,12 +78,12 @@ function BlockHooksControlPure( props ) {
...
 							// as a hooked first or last child block, as the block might've been automatically
 							// inserted and then moved around a bit by the user.
-							candidates = getBlock( props.clientId ).innerBlocks;
+							candidates = getBlock( clientId ).innerBlocks;
 							break;
 					}
 
 					const hookedBlock = candidates?.find(
-						( { name } ) => name === block.name
+						( candidate ) => name === candidate.name
```
Comment (ockham) [inline packages/block-editor/src/hooks/block-hooks.js:86]: I just noticed that this broke the toggle 😬 

I've filed a fix: https://github.com/WordPress/gutenberg/pull/57956.

Comment (ellatrix) [reply packages/block-editor/src/hooks/block-hooks.js:86]: Whoops, sorry!
